### PR TITLE
LDAP analyzer

### PR DESCRIPTION
**Pull request by iunanua**
### What does this PR do?
Instrument `ldapjs` library and add LDAP injection vulnerability detection

### Plugin Checklist

- [ ] Unit tests.
- [ ] TypeScript [definitions][1].
- [ ] TypeScript [tests][2].
- [ ] API [documentation][3].
- [ ] CircleCI [jobs/workflows][4].
- [ ] Plugin is [exported][5].

[1]: https://github.com/DataDog/dd-trace-js/blob/master/index.d.ts
[2]: https://github.com/DataDog/dd-trace-js/blob/master/docs/test.ts
[3]: https://github.com/DataDog/dd-trace-js/blob/master/docs/API.md
[4]: https://github.com/DataDog/dd-trace-js/blob/master/.circleci/config.yml
[5]: https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/plugins/index.js

### CODE REVIEW

**Comment by tlhunter** [comment]
Can you also toss an entry into `docs/API.md` and `index.d.ts`?

**Review by rochdev — APPROVED** [review]

File: docs/API.md [hunk 1]
```diff
@@ -121,6 +122,7 @@ tracer.use('pg', {
 * [kafkajs](./interfaces/plugins.kafkajs.html)
 * [knex](./interfaces/plugins.knex.html)
 * [koa](./interfaces/plugins.koa.html)
+* [ldapjs](./interfaces/plugins.ldapjs.html)
```
Comment (uurien) [inline docs/API.md:125]: Why? We haven't create any plugin for ldapjs, we have created just a vulnerability analyzer

File: packages/datadog-instrumentations/src/ldapjs.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  })
+
+  shimmer.wrap(ldapjs.Client.prototype, '_send', _send => function () {
+    if (ldapSearchCh.hasSubscribers) {
+      const callbackIndex = getCallbackArgIndex(arguments)
+      if (callbackIndex > -1) {
+        const callback = arguments[callbackIndex]
+        arguments[callbackIndex] = shimmer.wrap(callback, function (err, corkedEmitter) {
+          if (typeof corkedEmitter === 'object' && typeof corkedEmitter['on'] === 'function') {
+            arguments[1] = shimmer.wrap(corkedEmitter, 'on', on => function (name, fn) {
```
Comment (uurien) [inline packages/datadog-instrumentations/src/ldapjs.js:51]: Is `arguments[1] = ` necessary?
```suggestion
            shimmer.wrap(corkedEmitter, 'on', on => function (name, fn) {
```

File: packages/datadog-instrumentations/src/ldapjs.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+              if (typeof fn === 'function') {
+                const callbackResource = new AsyncResource('bound-anonymous-fn')
+                arguments[1] = callbackResource.bind(fn)
```
Comment (uurien) [inline packages/datadog-instrumentations/src/ldapjs.js:54]: I understand that here, we are wrapping the method `callback` into `res.on('end', callback)`
But, is it possible to do `res.off('end', callback)`? If it is, we will send to `off` method different method that we are sending in the `on`, and it is not going to be unsubscribed.

File: packages/datadog-instrumentations/src/ldapjs.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  }
+  return callbackIndex
+}
+
+addHook({ name: 'ldapjs', versions: ['>=2'] }, ldapjs => {
+  const ldapSearchCh = channel('datadog:ldapjs:client:search')
+
+  shimmer.wrap(ldapjs.Client.prototype, 'search', search => function (base, options) {
+    if (ldapSearchCh.hasSubscribers) {
+      let filter = null
```
Comment (uurien) [inline packages/datadog-instrumentations/src/ldapjs.js:30]: Any reason to initialize to `null` instead of maintain as `undefined`?
```suggestion
      let filter
```

File: packages/dd-trace/src/appsec/iast/analyzers/vulnerability-analyzer.js [hunk 5]
```diff
@@ -43,6 +58,10 @@ class Analyzer extends Plugin {
     }
   }
 
+  analyzeAll (...values) {
+    values.forEach(value => this.analyze(value))
```
Comment (uurien) [inline packages/dd-trace/src/appsec/iast/analyzers/vulnerability-analyzer.js:62]: It is not optimal, because each `this.analyze` tries to get store, context etc. It think that it will be better if we implement custom `analyze` in `ldapjs-injection-analyzer` (like in path-traversal-analyzer) or if we create other `analyzeAll` method getting the context only once.

File: packages/dd-trace/test/appsec/iast/analyzers/ldap-injection-analyzer.ldapjs.plugin.spec.js [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+          })
+        }, 'LDAP_INJECTION')
+      })
+
+      describe('has no vulnerability', () => {
+        testThatRequestHasNoVulnerability(() => {
+          return new Promise((resolve, reject) => {
+            const filter = '(objectClass=*)'
+            client.search(base, filter, (err, searchRes) => {
+              searchRes.on('end', resolve)
```
Comment (uurien) [inline packages/dd-trace/test/appsec/iast/analyzers/ldap-injection-analyzer.ldapjs.plugin.spec.js:54]: if `searchRes.off` is possible, it will be nice to have other test subscribing and unsubscribing one method, and checking that the unsubscribed is not executed.

File: packages/datadog-instrumentations/src/ldapjs.js (hunk 2 above)
Comment (iunanua) [reply packages/datadog-instrumentations/src/ldapjs.js:51]: Yes, it could be simplified

File: packages/datadog-instrumentations/src/ldapjs.js (hunk 3 above)
Comment (iunanua) [reply packages/datadog-instrumentations/src/ldapjs.js:54]: The event API allows to remove the listener with `off` or `removeListener` methods, so you are right.
In any case, I think the LDAP client would not emit more than one times 'end' or 'error' events, so it should not affect us (even without wrapping `off`or `removeListener`) but it is better to be safe than sorry :)

File: packages/dd-trace/test/appsec/iast/analyzers/ldap-injection-analyzer.ldapjs.plugin.spec.js (hunk 6 above)
Comment (iunanua) [reply packages/dd-trace/test/appsec/iast/analyzers/ldap-injection-analyzer.ldapjs.plugin.spec.js:54]: Added

File: packages/datadog-instrumentations/src/ldapjs.js [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+  return callbackIndex
+}
+
+const emitters = new WeakMap()
+
+function wrapEmitter (corkedEmitter) {
+  shimmer.wrap(corkedEmitter, 'on', on => function (name, fn) {
+    if (typeof fn === 'function') {
+      const callbackResource = new AsyncResource('bound-anonymous-fn')
+      const bindedFn = callbackResource.bind(fn)
```
Comment (uurien) [inline packages/datadog-instrumentations/src/ldapjs.js:31]: I think that you shoud check before create the callback resource if it already exist in callbackMap. If someone does ```
res.on('error', fn)
res.on('error', fn)
```
we should add the same function twice.

File: packages/datadog-instrumentations/src/ldapjs.js [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+      break
+    }
+  }
+  return callbackIndex
+}
+
+const emitters = new WeakMap()
+
+function wrapEmitter (corkedEmitter) {
+  shimmer.wrap(corkedEmitter, 'on', on => function (name, fn) {
```
Comment (uurien) [inline packages/datadog-instrumentations/src/ldapjs.js:28]: if we can do `off` and `removeListener`,  we can do also `addListener`

File: packages/dd-trace/src/appsec/iast/analyzers/vulnerability-analyzer.js [hunk 9]
```diff
@@ -37,10 +58,16 @@ class Analyzer extends Plugin {
...
-      this._report(value, iastContext)
-    }
+    this._reportIfVulnerable(value, iastContext)
+  }
+
+  analyzeAll (...values) {
+    const iastContext = getIastContext(storage.getStore())
+    if (!iastContext) return
+
+    values.forEach(value => this._reportIfVulnerable(value, iastContext))
```
Comment (uurien) [inline packages/dd-trace/src/appsec/iast/analyzers/vulnerability-analyzer.js:70]: If this method has 2 vulnerabilities, we are going to send to the backend only 1 (same file and line), but we will waste time calculating the stack twice.

We should change `_reportIfVulnerable` method to return `true` when it is vulnerable, and break the for:

```suggestion
    for (let i = 0; i < values.length; i++) {
      if (this._reportIfVulnerable(values[i], iastContext)) break;
    }
```

Comment (iunanua) [reply packages/dd-trace/src/appsec/iast/analyzers/vulnerability-analyzer.js:70]: Then 'analyzeAll` method will assume that values are from the same file and line... 
Will the condition always be met?

Comment (uurien) [reply packages/dd-trace/src/appsec/iast/analyzers/vulnerability-analyzer.js]: Yes, the line and file are calculated in report method, so the stack is always the same.

File: packages/datadog-instrumentations/src/ldapjs.js [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+        }
+      }
+      ldapSearchCh.publish({ base, filter })
+    }
+
+    return search.apply(this, arguments)
+  })
+
+  shimmer.wrap(ldapjs.Client.prototype, '_send', _send => function () {
+    if (ldapSearchCh.hasSubscribers) {
```
Comment (uurien) [inline packages/datadog-instrumentations/src/ldapjs.js:76]: This code block is not related directly related with ldapjs injection, it is to prevent the context from being lost.

I think that we should do this even when IAST is not active and nothing is listening the channel.

File: packages/dd-trace/test/appsec/iast/analyzers/ldap-injection-analyzer.ldapjs.plugin.spec.js [hunk 11]
```diff
@@ -0,0 +1,113 @@
+'use strict'
+
+const { testThatRequestHasVulnerability, testThatRequestHasNoVulnerability } = require('../utils')
+const { storage } = require('../../../../../datadog-core')
+const iastContextFunctions = require('../../../../src/appsec/iast/iast-context')
+const { newTaintedString } = require('../../../../src/appsec/iast/taint-tracking/operations')
+const vulnerabilityReporter = require('../../../../src/appsec/iast/vulnerability-reporter')
+const agent = require('../../../plugins/agent')
+const { emit } = require('process')
```
Comment (uurien) [inline packages/dd-trace/test/appsec/iast/analyzers/ldap-injection-analyzer.ldapjs.plugin.spec.js:9]: linter fails

**Review by uurien — APPROVED** [review]

**Comment by iunanua** [comment]
:tada: :tada: :tada:
